pkg/controllers/metrics/pod/controller.go

```diff
@@ -0,0 +1,186 @@
+	"github.com/prometheus/client_golang/prometheus"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	podName             = "name"
+	podNameSpace        = "namespace"
+	ownerSelfLink       = "owner"
+	podHostName         = "node"
+	podProvisioner      = "provisioner"
+	podHostZone         = "zone"
+	podHostArchitecture = "arch"
+	podHostCapacityType = "capacity_type"
+	podHostInstanceType = "instance_type"
+	podPhase            = "phase"
+)
+
+var (
+	podGaugeVec = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "karpenter",
+			Subsystem: "pods",
+			Name:      "state",
+			Help:      "Pod state.",
+		},
+		labelNames(),
+	)
+)
+
+// Controller for the resource
+type Controller struct {
+	KubeClient client.Client
+	LabelsMap  map[types.NamespacedName]prometheus.Labels
+}
+
+func init() {
+	crmetrics.Registry.MustRegister(podGaugeVec)
+}
+
+func labelNames() []string {
+	return []string{
+		podName,
+		podNameSpace,
+		ownerSelfLink,
+		podHostName,
+		podProvisioner,
+		podHostZone,
+		podHostArchitecture,
+		podHostCapacityType,
+		podHostInstanceType,
+		podPhase,
+	}
+
+}
+
+// NewController constructs a controller instance
+func NewController(kubeClient client.Client) *Controller {
+	return &Controller{
+		KubeClient: kubeClient,
+		LabelsMap:  make(map[types.NamespacedName]prometheus.Labels),
+	}
+}
+
+// Reconcile executes a termination control loop for the resource
+func (c *Controller) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("podmetrics").With("pod", req.Name))
+	// Retrieve pod from reconcile request
+	pod := &v1.Pod{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, pod); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to pod deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				podGaugeVec.Delete(labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge, failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after pod labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		podGaugeVec.Delete(labels)
+	}
+	if err := c.updateGauges(ctx, pod); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to record metrics, %s", err.Error())
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) updateGauges(ctx context.Context, pod *v1.Pod) error {
+	newlabels := c.generateLabels(ctx, pod)
+	gauge, err := podGaugeVec.GetMetricWith(newlabels)
```

Comment:
We control all the inputs and outputs here, so we can guarantee it won't panic if you use `With` instead:
```
// With works as GetMetricWith, but panics where GetMetricWithLabels would have
// returned an error. Not returning an error allows shortcuts like
//     myVec.With(prometheus.Labels{"code": "404", "method": "GET"}).Add(42)
func (v *GaugeVec) With(labels Labels) Gauge {
	g, err := v.GetMetricWith(labels)
	if err != nil {
		panic(err)
	}
	return g
}
```

Reply:
This simplifies your implementation to
```
podGaugeVec.With(labels).Set(float64(1))
c.LabelsMap[client.ObjectKeyFromObject(pod)] labels
```

Reply:
Great suggestion! done!